evm/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -565,15 +565,29 @@ where
             }
         } else if call.contract != h160_to_b160(HARDHAT_CONSOLE_ADDRESS) {
             // Handle expected calls
-            if let Some(expecteds) = self.expected_calls.get_mut(&(b160_to_h160(call.contract))) {
-                if let Some((_, count)) = expecteds.iter_mut().find(|(expected, _)| {
-                    expected.calldata.len() <= call.input.len() &&
-                        expected.calldata == call.input[..expected.calldata.len()] &&
-                        expected.value.map_or(true, |value| value == call.transfer.value.into()) &&
+
+            // Grab the different calldatas expected.
+            if let Some(expected_calls_for_target) =
+                self.expected_calls.get_mut(&(b160_to_h160(call.contract)))
+            {
+                // Match every partial/full calldata
+                for (calldata, (expected, actual_count)) in expected_calls_for_target.iter_mut() {
+                    // Increment actual times seen if...
+                    // The calldata is at most, as big as this call's input, and
+                    if calldata.len() <= call.input.len() &&
+                        // Both calldata match, taking the length of the assumed smaller one (which will have at least the selector), and
+                        *calldata == call.input[..calldata.len()] &&
+                        // The value matches, if provided
+                        expected
+                            .value
+                            .map_or(true, |value| value == call.transfer.value.into()) &&
+                        // The gas matches, if provided
                         expected.gas.map_or(true, |gas| gas == call.gas_limit) &&
+                        // The minimum gas matches, if provided
                         expected.min_gas.map_or(true, |min_gas| min_gas <= call.gas_limit)
-                }) {
-                    *count += 1;
+                    {
+                        *actual_count += 1;
```

Comment:
For any of the two modes, the discrimination of how to interpret the actual count is done later—our only job here is detecting if all values properly match for the current call.